[WIP, RFC] Introduce Gy interface

pespin opened the pull request:
TODO:

* Use an event for the report, like SMF_EVT_N4_TIMER?

*  Properly set Service identifier in Gy CCR

* SMF: Properly set pkt/octet volumes in Gy CCR
** Update when receiving PFCP Modify Response.

* Figure out best way to require Gy through config file in open5gs-smfd.

* Create a new sess-sm.c which handles smf_sess_t state through
  Gx+Gy+PFCP creation, modification and tear down. This way we can do
stuff in parallel, for instance Gx+Gy. It will alsoavoid duplicating
some code paths due to Gy being optional.

pespin commented:
Hi @acetcom ,

Please don't merge yet this PR, it's still not 100% ready. I submitted it here now in order to get early review from you. Feel free to provide feedback or advise on any topic.
See in the PR description some of the TODO things.

acetcom commented:
@pespin 

Since starting this project, I've never seen anyone handle Open5GS as much as you do.

Basically, I couldn't find any problems with the overall architecture. 100% agree with your design.


I will only mention a few very minor parts.

1. Instead of stopping the timer, it would be better to delete the timer.
```
$ diff --git a/lib/pfcp/context.c b/lib/pfcp/context.c
index 08b389d30..862b1a8f7 100644
--- a/lib/pfcp/context.c
+++ b/lib/pfcp/context.c
@@ -1311,7 +1311,7 @@ void ogs_pfcp_urr_remove(ogs_pfcp_urr_t *urr)
     ogs_list_remove(&sess->urr_list, urr);
 
     if (urr->t_time_threshold)
-        ogs_timer_stop(urr->t_time_threshold);
+        ogs_timer_delete(urr->t_time_threshold);
 
     if (urr->id_node)
         ogs_pool_free(&urr->sess->urr_id_pool, urr->id_node);
```

2. Fix Indentation
```
$ diff --git a/src/smf/smf-sm.c b/src/smf/smf-sm.c
index 2af764da6..dc007c9fc 100644
--- a/src/smf/smf-sm.c
+++ b/src/smf/smf-sm.c
@@ -272,48 +272,48 @@ void smf_state_operational(ogs_fsm_t *s, smf_event_t *e)
         ogs_free(gx_message);
         break;
 
-        case SMF_EVT_GY_MESSAGE:
-            ogs_assert(e);
-            gy_message = e->gy_message;
-            ogs_assert(gy_message);
+    case SMF_EVT_GY_MESSAGE:
+        ogs_assert(e);
+        gy_message = e->gy_message;
+        ogs_assert(gy_message);
 
-            sess = e->sess;
-            ogs_assert(sess);
+        sess = e->sess;
+        ogs_assert(sess);
 
-            switch(gy_message->cmd_code) {
-            case OGS_DIAM_GY_CMD_CODE_CREDIT_CONTROL:
-                switch(gy_message->cc_request_type) {
-                case OGS_DIAM_GY_CC_REQUEST_TYPE_INITIAL_REQUEST:
-                    ogs_assert(e->gtp_xact);
-                    smf_gy_handle_cca_initial_request(
-                            sess, gy_message, e->gtp_xact);
-                    break;
-                case OGS_DIAM_GY_CC_REQUEST_TYPE_UPDATE_REQUEST:
-                        ogs_assert(e->pfcp_xact);
-                        smf_gy_handle_cca_update_request(
-                                sess, gy_message, e->pfcp_xact);
+        switch(gy_message->cmd_code) {
+        case OGS_DIAM_GY_CMD_CODE_CREDIT_CONTROL:
+            switch(gy_message->cc_request_type) {
+            case OGS_DIAM_GY_CC_REQUEST_TYPE_INITIAL_REQUEST:
+                ogs_assert(e->gtp_xact);
+                smf_gy_handle_cca_initial_request(
+                        sess, gy_message, e->gtp_xact);
                 break;
-                case OGS_DIAM_GY_CC_REQUEST_TYPE_TERMINATION_REQUEST:
-                    ogs_assert(e->gtp_xact);
-                    smf_gy_handle_cca_termination_request(
-                            sess, gy_message, e->gtp_xact);
-                    break;
-                default:
-                    ogs_error("Not implemented(%d)", gy_message->cc_request_typ:
- e);
-                    break;
-                }
-
+            case OGS_DIAM_GY_CC_REQUEST_TYPE_UPDATE_REQUEST:
+                ogs_assert(e->pfcp_xact);
+                smf_gy_handle_cca_update_request(
+                        sess, gy_message, e->pfcp_xact);
                 break;
-            case OGS_DIAM_GY_CMD_RE_AUTH:
-                smf_gy_handle_re_auth_request(sess, gy_message);
+            case OGS_DIAM_GY_CC_REQUEST_TYPE_TERMINATION_REQUEST:
+                ogs_assert(e->gtp_xact);
+                smf_gy_handle_cca_termination_request(
+                        sess, gy_message, e->gtp_xact);
                 break;
             default:
-                ogs_error("Invalid type(%d)", gy_message->cmd_code);
+                ogs_error("Not implemented(%d)", gy_message->cc_request_type);
                 break;
             }
 
-            ogs_free(gy_message);
             break;
+        case OGS_DIAM_GY_CMD_RE_AUTH:
+            smf_gy_handle_re_auth_request(sess, gy_message);
+            break;
+        default:
+            ogs_error("Invalid type(%d)", gy_message->cmd_code);
+            break;
+        }
+
+        ogs_free(gy_message);
+        break;
```

3. How about remove union in the smf/event.h?

You can leave it alone, but other events(MME/SGW/..) don't use it, so I just mentioned it.
```
$ diff --git a/src/smf/event.h b/src/smf/event.h
index 8b6cfca30..d3f70c4a6 100644
--- a/src/smf/event.h
+++ b/src/smf/event.h
@@ -81,10 +81,8 @@ typedef struct smf_event_s {
     ogs_pfcp_xact_t *pfcp_xact;
     ogs_pfcp_message_t *pfcp_message;
 
-    union {
-        ogs_diam_gx_message_t *gx_message;
-        ogs_diam_gy_message_t *gy_message;
-    };
+    ogs_diam_gx_message_t *gx_message;
+    ogs_diam_gy_message_t *gy_message;
 
     struct {
         ogs_sbi_request_t *request;

```

Let me know when you're ready. I'll mrege this pull request at that time.

Thank you so much!
Sukchan

pespin commented:
Hi @acetcom ,

I submitted several other PRs with stuff which is ready to be merged.
I am breaking this current PR into several commits which are a bit easier to digest. I am also fixing some stuff from your last review.

I have several comments still which need discussion/advise from you:

> 3. How about remove union in the smf/event.h?

I think it makes sense to use a union there, I was actually surprised that none was used before. When an event is sent, it usually only uses one of the fields, and never both, so having them in a union:
* Makes it easier to understand to the reader than the 2 fields cannot be used at the same time
* Make allocation size for each event smaller.

Still if you don't like to use unions there I can of course drop it.

> Figure out best way to require Gy through config file in open5gs-smfd.

I consider the Gy interface to be optional, and open5gs-smfd should require an OCS being connected to work properly. This is also important since there's no open5gs-ocs yet.
In any case, it can make sense for a commercial operator to *require* the availability of an OCS to accept user sessions. Furthermore, an operator may want to ignore the fact that there's a Diameter peer announcing Gy support and make open5gs-smfd don't use it.

As a result, I think it makes sense to have a tri-state configuration option in open5gs-smfd.yml, something like:
```
ctf:
  use: auto
```
Where "ctf" relates tothe CTF node, the one connecting to OCS and which open5gs-smfd implements.
Attribute "use" can take values "auto|yes|no":
* "auto": Use Gy if a diameter peer (OCS) announcing support for it is connected `ogs_diam_app_connected(OGS_DIAM_GY_APPLICATION_ID)`
* "yes": Force Gy use. If no Gy diameter peer (OCS) is available, reject sessions
* "no": Don't use Gy even if there's a GY diameter peer (OCS) available

The default option would be "auto".

We can later on add more config options under ctf, like threshold calculation based on quotas, etc.

Do you agree with my proposal?


PS: Is there a way to change configuration options without restarting open5gs programs? Something similar to osmocom VTY. That would be great to have so we can use it in our TTCN3 tests. It would also be great for operators I guess.

acetcom commented:
I agree that the `ctf` option is a good idea. Also, to TTCN3 testing with the `ctf` option, it seems like I need to be able to change the configuration without restarting the program. Unfortunately, I can't do this part right now because I have other work to do.

`union` doesn't mean much. I just did it to reduce mistakes. I think it's okay to use union unless gx/gy is always used at the same time. Whichever one you choose is fine.

BTW, I noticed that you have split your work into multiple pull requests. Can I merge these now? Or should I wait?

Thank you so much for your contribution!
Sukchan

pespin commented:
> BTW, I noticed that you have split your work into multiple pull requests. Can I merge these now? Or should I wait?

Yes please, merge all the other PRs (not this one) ASAP.

> Also, to TTCN3 testing with the ctf option, it seems like I need to be able to change the configuration without restarting the program.

I think using "auto" in this case would be enough, because I can emulate the OCS only in the tests I want Gy to occur. But yes, in general I think having some sort of "reload/change/introspect" configuration is a major thing open5gs lacks.

pespin commented:
I'll let you know when this PR is ready. Now I just rebased the remaining WIP patch on top of current main.

pespin commented:
You should have not merged this PR yet, as I stated when I updated it:

> I'll let you know when this PR is ready. Now I just rebased the remaining WIP patch on top of current main.

Please revert it, it will break use for other people.

acetcom commented:
@pespin 

Oops! I misunderstood your intention. So, I reversed this PR.
reverted

Let me know if this PR is ready again!
Sukchan
